### PR TITLE
fix issue 12897 - toJSON, add the escapeNonAsciiChars option

### PR DESCRIPTION
**Pull request by BBasile**
This new option allows a better interoperability with software made in other languages and that are not able to read string values containing non-ASCII characters.

- no breakage
- no additional code (excepted a test) because the encoder already existed for unicode ctrl chars

### CODE REVIEW

File: std/json.d [hunk 1]
```diff
@@ -1171,12 +1172,12 @@ string toJSON(in JSONValue* root, in bool pretty = false, in JSONOptions options
...
 
-private void appendJSONChar(Appender!string* dst, dchar c,
-                            scope void delegate(string) error)
+private void appendJSONChar(Appender!string* dst, dchar c, in JSONOptions options,
+    scope void delegate(string) error)
 {
     import std.uni : isControl;
 
-    if(isControl(c))
+    with (JSONOptions) if (isControl(c) || ((options & escapeMBC) >= escapeMBC && c >= 0x80))
```
Comment (quickfur) [inline std/json.d]: Nitpick: this line seems a little long; wrap it?

File: std/json.d [hunk 2]
```diff
@@ -73,6 +73,7 @@ Flags that control how json is encoded and parsed.
 enum JSONOptions {
     none,                       /// standard parsing
     specialFloatLiterals = 0x1, /// encode NaN and Inf float values as strings
+    escapeMBC = 0x2             /// encode multi byte characters using an unicode escape sequence
```
Comment (quickfur) [inline std/json.d]: I'm not sure I like the name `escapeMBC`... Maybe `escapeMultibyteChar`? I know it's a bit more verbose, but the acronym MBC made no sense to me until I read the docs.

File: std/json.d [hunk 3]
```diff
@@ -1025,11 +1027,10 @@ string toJSON(in JSONValue* root, in bool pretty = false, in JSONOptions options
                 case '\r':      json.put("\\r");        break;
                 case '\t':      json.put("\\t");        break;
                 default:
-                    appendJSONChar(&json, c,
-                                   (msg) { throw new JSONException(msg); });
+                    appendJSONChar(&json, c, options,
+                        (msg) { throw new JSONException(msg); });
```
Comment (quickfur) [inline std/json.d]: Nitpick: lining this line up with the opening `(` (like it was originally) looks nicer. Not sure why the indent was changed when all that was needed was to append `options` to the end of the previous line.

**Comment by quickfur** [comment]
Other than that, LGTM.

**Comment by BBasile** [comment]
_escapeNonAsciiChars_ is much more accurate

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by BBasile** [comment]
conflicts inducted by #4108 solved and rebased.

**Comment by BBasile** [comment]
@braddr: you probably have noticed this too, win-farm-1 prevents PRs to be merged since almost 24 hours with always an OutOfMemory error. While this one initially had to be rebased it's more clear if you look at the [last column](https://auto-tester.puremagic.com/pulls.ghtml?projectid=1). Even in the [GH PR list](https://github.com/D-Programming-Language/phobos/pulls) there's no green check marks anymore, which is quite abnormal.

**Comment by braddr** [comment]
Yes, it is.  Not sure why you blame the host or auto-tester rather than 
dmd, a known memory hog.  Something's changed and it now frequently runs 
out of memory.  Someone needs to spend some time diagnosing what changed 
in the last few weeks and/or directly working on the memory overhead of dmd.

**Comment by BBasile** [comment]
If it only happens on a 32 bit machine that generates 64 bit code that's already a trail.

**Comment by schveiguy** [comment]
@BBasile It's not a 32-bit machine, just a 32-bit compiler that generates 64-bit code. You wouldn't be able to build if it was a 32-bit machine I don't think.

@braddr I imagine what has happened is that the memory usage has crept to the threshold for dmd. I'm not sure what we can do about it, IIRC @WalterBright, the compiler never frees memory in exchange for compilation speed. It's likely we have to figure out a better way. Not completing is worse than completing slower.

**Comment by schveiguy** [comment]
In addition, I think win farm 1 is the only 64-bit windows tester. So we have a big problem if this can't pass anything.

**Comment by BBasile** [comment]
There's definitively a bug in the auto integration process. It's still marked for merging on next successful test battery but now that it's happened, it's not merged.

Maybe the reason is that the Dlang-PL member who has toogled auto-merge is undefined: (see https://auto-tester.puremagic.com/pull-history.ghtml?projectid=1&repoid=3&pullid=4106) __Auto-Merge	Yes - (by ) – Toggle__.

**Comment by BBasile** [comment]
@braddr , can you look at https://auto-tester.puremagic.com/pull-history.ghtml?projectid=1&repoid=3&pullid=4106 ? I don't know if you've get the notification for my previous online message but it seems that there's an auto-tester bug.

- non member PR author
- toogled for merge by member
- closed by author
- reopened by author
- still marked for mergeing but merge doesn't happen when all hosts have passed.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled off

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by schveiguy** [comment]
@BBasile Seems the auto tester lost track of who was supposed to auto-merge it. It needs github credentials to do so.

**Pull request merged**